src/ILCompiler.Compiler/src/CppCodeGen/ILToCppImporter.cs

```diff
@@ -2249,11 +2266,14 @@ private void ImportLdToken(int token)
                     "((intptr_t)",
                     _writer.GetCppTypeName((TypeDesc)ldtokenValue),
                     "::__getMethodTable())");
+
+                value = new MetaDataEntry<TypeDesc>(StackValueKind.ValueType, name, (TypeDesc)ldtokenValue, GetWellKnownType(ldtokenKind));
             }
             else if (ldtokenValue is FieldDesc)
             {
                 ldtokenKind = WellKnownType.RuntimeFieldHandle;
                 name = null;
+                value = new MetaDataEntry<TypeSystem.Ecma.EcmaField>(StackValueKind.ValueType, name, (TypeSystem.Ecma.EcmaField) ldtokenValue, GetWellKnownType(ldtokenKind));
```

Comment:
This should be `MetaDataEntry<FieldDesc>`. The cast to `EcmaFieldDesc` should be done in the one place where it is needed to the array initialization.